resource/traversers/dfu.cpp

```
@@ -288,29 +290,47 @@ int dfu_traverser_t::run (Jobspec::Jobspec &jobspec,
     }
 
     int rc = -1;
+    int64_t graph_end = graph_duration.graph_end.time_since_epoch ().count ();
     detail::jobmeta_t meta;
     vtx_t root = get_graph_db ()->metadata.roots.at (dom);
     bool x = detail::dfu_impl_t::exclusivity (jobspec.resources, root);
     const std::set<std::string> exclusive_types =
                         detail::dfu_impl_t::get_exclusive_resource_types ();
     std::unordered_map<std::string, int64_t> dfv;
+
     detail::dfu_impl_t::prime_jobspec (jobspec.resources, dfv);
-    if (meta.build (jobspec,
-                    detail::jobmeta_t::alloc_type_t::AT_ALLOC, jobid, *at) < 0)
+    if (meta.build (jobspec, detail::jobmeta_t::alloc_type_t::AT_ALLOC, jobid,
+                    *at, graph_duration) < 0)
         return -1;
 
     if ( (op == match_op_t::MATCH_SATISFIABILITY)
         && (rc = is_satisfiable (jobspec, meta, x, root, dfv)) == 0) {
         detail::dfu_impl_t::update ();
     } else if ( (rc = schedule (jobspec, meta, x, op, root, dfv)) ==  0) {
         *at = meta.at;
+        // First check that meta.duration is shorter than expressable 
+        // int64_t max ()
+        if (meta.duration > std::numeric_limits<int64_t>::max ())
+            goto error;
+        if (*at < 0 or *at >= graph_end)
+            goto error;
+        // If job ends after the resource graph expires, reduce the duration
+        // so it coincides with the graph expiration. Note that we could
+        // arguably return ENOTSUP/EINVAL instead. Also note that we 
+        // know *at < graph_end from the previous check.
+        if ( (*at + static_cast<int64_t> (meta.duration)) > graph_end)
+            meta.duration = graph_end - *at;
```

@grondo @garlick Dan and I wanted to see what you think of this (see comment and above two lines of code). It would mean that more jobs would run, but not always for the duration that users requested. E.g. if the scheduler has 10 minutes left before resource expiration and a 30 minute job comes in, the job might be run (rather than rejected) but with a 10 minute duration.

---

One more thing we should address along these lines. The current check for jobspec duration here: https://github.com/flux-framework/flux-sched/blob/6483346a3ecb9b7bccb103d760a0e87f92df6063/resource/traversers/dfu_impl.hpp#L73
does effectively the same thing in that it shortens the job duration to `SYSTEM_MAX_DURATION` (Fluxion master) or the graph duration (this PR). Perhaps it should return an error, too?

---

> if the scheduler has 10 minutes left before resource expiration and a 30 minute job comes in, the job might be run (rather than rejected) but with a 10 minute duration.

This seems like a fine compromise. I should note that we do have the `duration-validator` jobtap plugin in flux-core, which will reject jobs with a duration longer than the remaining instance lifetime, so you won't have a job _submitted_ that could exceed the instance lifetime. However, you might have a job that fits within the instance expiration at the time of submission, but at the time resources are allocated it no longer does. In that case it makes total sense to reduce the expiration of the job to the expiration of the instance, but still let the job run.

---

There are other use cases for Fluxion (like Kubernetes integrations) where the inputs probably won't be validated by the submitter.

> However, you might have a job that fits within the instance expiration at the time of submission, but at the time resources are allocated it no longer does. 

According to the scheduler the resources are allocated immediately after the code block in this comment thread, so I don't think that will happen in practice. 

I'm ambivalent about this approach because it seems like the user will expect that if they submit a job with a specified duration, the duration will either be honored or rejected. Getting the duration truncated might be confusing.

---

@trws, what are your thoughts?

---

I'm missing a bit of context, if we're talking about submission time then I would probably prefer that the job be rejected.  If we're talking about processing the job later, maybe after a re-schedule where a higher priority job bumped the current one later, then giving it a chance to at least start sounds sensible.

---

> >     However, you might have a job that fits within the instance expiration at the time of submission, but at the time resources are allocated it no longer does.

> 
> According to the scheduler the resources are allocated immediately after the code block in this comment thread, so I don't think that will happen in practice.

Isn't that the problem though? A job is submitted to flux with duration T and is accepted by flux-core because the instance will not expire for another T + 5 seconds, but then it takes 10 seconds for the job to make it to the point of being scheduled, at which point the duration cannot be honored.

In any case I think it might not be a bad idea to ask Ryan and go with whatever he suggests. It's not really a system instance issue, but still.

---

I agree, better to let the job run and get some work done.  And I think that would work well with what libyogrt would expect.

I was going to say we would want to error out if the duration truncation were large, as a feedback mechanism.  But it seems like that means the scheduling delay is large, so either something is broken or the node is overloaded, and it would be better for the user/admin to learn about that in a way that does not prevent jobs from running.

That said, Ryan likely has a much better idea of what users expect than I do.

---

There seems to be a consensus and I'm not questioning that but for my own edification:

> A job is submitted to flux with duration T and is accepted by flux-core because the instance will not expire for another T + 5 seconds, but then it takes 10 seconds for the job to make it to the point of being scheduled, at which point the duration cannot be honored because the instance expires in T - 5 seconds.

When we say "scheduled" I assume that means "future start time planned" not "allocated"?  (Since obviously we expect jobs to be pending for long periods of time in a batch environment)

> But it seems like that means the scheduling delay is large, so either something is broken or the node is overloaded, and it would be better for the user/admin to learn about that in a way that does not prevent jobs from running.

Does fluxion's configured "queue-depth" still potentially mean that some pending jobs may remain "off the plan" for a long period after submission?  (The default queue depth appears to be [quite low](https://github.com/flux-framework/flux-sched/blob/409de4e9f32593e59603bdbfcd1809f4d88084b6/qmanager/config/queue_system_defaults.hpp#L17) so hopefully I'm misinterpreting what that means)

---

> When we say "scheduled" I assume that means "future start time planned" not "allocated"? (Since obviously we expect jobs to be pending for long periods of time in a batch environment)

Well technically by "scheduled" I meant "the job hits this check of ours that shortens the duration", rather ignorant of where exactly in the job lifecycle this check occurs. But yes I think that implies "future start time planned" rather than allocated. I have no idea how many jobs remain "off the plan" for long periods (maybe Dan knows), I was just thinking that it's not zero.

---

> When we say "scheduled" I assume that means "future start time planned" not "allocated"? (Since obviously we expect jobs to be pending for long periods of time in a batch environment)

I think either possibility is valid given what I know about the job lifecycle. 

> Does fluxion's configured "queue-depth" still potentially mean that some pending jobs may remain "off the plan" for a long period after submission?

I think this is possible given the queue depth, but I confess `qmanager` isn't a component I know much about yet.

Given your comments @garlick and @ofaaland I think shortening the duration is the right approach.